swarm/network_test.go

```diff
@@ -0,0 +1,460 @@
+	// uniqueness is very certain.
+	data := fmt.Sprintf("test content %s %x", time.Now().Round(0), b)
+	k, wait, err := swarm.api.Put(data, "text/plain", false)
+	if err != nil {
+		return nil, "", err
+	}
+	if wait != nil {
+		wait()
+	}
+	return k, data, nil
+}
+
+func retrieve(net *simulations.Network, files []file, swarms map[discover.NodeID]*Swarm, trigger chan discover.NodeID) (missing uint64) {
+	var checkStatusM sync.Map
+	var nodeStatusM sync.Map
+	var totalFoundCount uint64
+
+	rand.Shuffle(len(files), func(i, j int) {
+		files[i], files[j] = files[j], files[i]
+	})
+
+	var totalWg sync.WaitGroup
+	errc := make(chan error)
+
+	nodeIDs := allNodeIDs(net)
+
+	totalCheckCount := len(nodeIDs) * len(files)
+
+	for _, id := range nodeIDs {
+		if _, ok := nodeStatusM.Load(id); ok {
+			continue
+		}
+		start := time.Now()
+		var checkCount uint64
+		var foundCount uint64
+
+		totalWg.Add(1)
+
+		var wg sync.WaitGroup
+
+		for _, f := range files {
+			swarm := swarms[id]
+
+			checkKey := check{
+				key:    f.key.String(),
+				nodeID: id,
+			}
+			if n, ok := checkStatusM.Load(checkKey); ok && n.(int) == 0 {
+				continue
+			}
+
+			checkCount++
+			wg.Add(1)
+			go func(f file, id discover.NodeID) {
+				defer wg.Done()
+
+				log.Debug("api get: check file", "node", id.String(), "key", f.key.String(), "total files found", atomic.LoadUint64(&totalFoundCount))
+
+				r, _, _, err := swarm.api.Get(f.key, "/")
+				if err != nil {
+					errc <- fmt.Errorf("api get: node %s, key %s, kademlia %s: %v", id, f.key, swarm.bzz.Hive, err)
+					return
+				}
+				d, err := ioutil.ReadAll(r)
+				if err != nil {
+					errc <- fmt.Errorf("api get: read response: node %s, key %s: kademlia %s: %v", id, f.key, swarm.bzz.Hive, err)
+					return
+				}
+				data := string(d)
+				if data != f.data {
+					errc <- fmt.Errorf("file contend missmatch: node %s, key %s, expected %q, got %q", id, f.key, f.data, data)
+					return
+				}
+				checkStatusM.Store(checkKey, 0)
+				atomic.AddUint64(&foundCount, 1)
+				log.Info("api get: file found", "node", id.String(), "key", f.key.String(), "content", data, "files found", atomic.LoadUint64(&foundCount))
+			}(f, id)
+		}
+
+		go func(id discover.NodeID) {
+			defer totalWg.Done()
+			wg.Wait()
+
+			atomic.AddUint64(&totalFoundCount, foundCount)
+
+			if foundCount == checkCount {
+				log.Info("all files are found for node", "id", id.String(), "duration", time.Since(start))
+				nodeStatusM.Store(id, 0)
+				trigger <- id
+				return
+			}
+			log.Debug("files missing for node", "id", id.String(), "check", checkCount, "found", foundCount)
+		}(id)
+
+	}
+
+	go func() {
+		totalWg.Wait()
+		close(errc)
+	}()
+
+	var errCount int
+	for err := range errc {
+		if err != nil {
+			errCount++
+		}
+		log.Error(err.Error())
+	}
+
+	log.Info("check stats", "total check count", totalCheckCount, "total files found", atomic.LoadUint64(&totalFoundCount), "total errors", errCount)
```

Comment:
why need atomic at this point?

Reply:
I had some bad value repostings without it. Maybe atomic can be replaced with a locked int as a counter. That would be more idiomatic and maybe even correct?